### PR TITLE
tests: Wrap or remove deprecated functions

### PR DESCRIPTION
**Pull request by pdgendt**
Make deprecated calls explicit during testing or remove them if unnecessary.

This allows us to promote warnings to errors with `pytest`.

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -29,10 +28,22 @@ def autouse_config_tmpdir(config_tmpdir):
...
 def cfg(f=ALL, topdir=None):
     # Load a fresh configuration object at the given level, and return it.
-    cp = configparser.ConfigParser(allow_no_value=True)
-    config.read_config(configfile=f, config=cp, topdir=topdir)
+    cp = config.configparser.ConfigParser(allow_no_value=True)
+    cp.read(config._gather_configs(f, topdir), encoding='utf-8')
     return cp
 
+def update_config(*args, **kwargs):
+    # For testing purposes it's easiest to use the deprecated functions
```
Comment (marc-hb) [inline tests/test_config.py]: Nit
```suggestion
    # For testing purposes it's easier to use the deprecated functions
```

File: tests/test_config.py [hunk 2]
```diff
@@ -29,10 +28,22 @@ def autouse_config_tmpdir(config_tmpdir):
 
 def cfg(f=ALL, topdir=None):
     # Load a fresh configuration object at the given level, and return it.
-    cp = configparser.ConfigParser(allow_no_value=True)
-    config.read_config(configfile=f, config=cp, topdir=topdir)
+    cp = config.configparser.ConfigParser(allow_no_value=True)
```
Comment (marc-hb) [inline tests/test_config.py:33]: Mostly not your fault but this commit is incredibly hard to review... first, this import rename is horrible because it hides both the origin and nature of `config`:
```python
from west import configuration as config
```
(don't change it now because it would have a massive ripple effect)

Then, I can't find `west.configuration.configparser` in the documentation. Is this is some "private" object (ab)used for "white box" testing? If yes please add a comment stating that. If not, sorry I missed what it is.

The icing on the cake is the fact that this commit replaces Python's built-in `configparser` with `west.configuration.configparser` which has the very same name while being a bit of a mystery...

I hate to ask but... could you slow down a little bit? Maybe move this to a separate commit. Either way, provide a bit more comments, commit message, etc. Thanks in advance!

File: tests/test_config.py [hunk 3]
```diff
@@ -29,10 +28,22 @@ def autouse_config_tmpdir(config_tmpdir):
...
+    cp.read(config._gather_configs(f, topdir), encoding='utf-8')
     return cp
 
+def update_config(*args, **kwargs):
+    # For testing purposes it's easiest to use the deprecated functions
+    # but make these calls explicit.
+    with pytest.deprecated_call():
+        return config.update_config(*args, **kwargs)
```
Comment (marc-hb) [inline tests/test_config.py]: Nice.

File: tests/test_project.py [hunk 4]
```diff
@@ -1631,7 +1630,6 @@ def test_init_local_with_manifest_filename(repos_tmpdir):
     # success
     cmd(['init', '--mf', 'project.yml', '-l', zephyr_install_dir])
     workspace.chdir()
-    config.read_config()
```
Comment (marc-hb) [inline tests/test_project.py:1634]: Should this be a separate commit?

File: tox.ini [hunk 5]
```diff
@@ -37,6 +37,6 @@ setenv =
     # For instance: ./.tox/py3/tmp/
     TOXTEMPDIR={envtmpdir}
 commands =
-  python -m pytest --cov-report=html --cov=west {posargs:tests} --basetemp='{envtmpdir}/pytest with space/'
+  python -m pytest -W error --cov-report=html --cov=west {posargs:tests} --basetemp='{envtmpdir}/pytest with space/'
```
Comment (marc-hb) [inline tox.ini:40]: Nice, but could there be a convenient way/setting to temporarily turn this off? I mean without `git diff` and `git status` pollution.

When hacking, there is often a intermediate stage where you really want the tests to pass on work in progress that still has warnings.

File: tests/test_config.py [hunk 6]
```diff
@@ -29,10 +28,22 @@ def autouse_config_tmpdir(config_tmpdir):
...
+
+def delete_config(*args, **kwargs):
```
Comment (marc-hb) [inline tests/test_config.py]: Warning: this review comment will be hard to find because that line  is changed in a later commit in the PR and Github doesn't like the multi-commit, force-push workflow.

There is already an API named `delete_config()`, so please use a different name for internal symbols. Example:
```suggestion
def delete_testcfg(*args, **kwargs):
```

Same for `update_config` above.

Sincere condolences for the ripple effect on consecutive commits :-(

Generally speaking, please choose less "API-looking" names for internal symbols. Pretty much everything here is called `something_config` and it's incredibly hard to keep track of what is: 1) Builtin Python (e.g. `configparser`) 2) deprecated west API 3) current west API 4) private west objects abused for "white box" testing 5) local/private test symbols,....

It's never going to be perfect and please do not cause massive git churn but brand new symbols is where progress can and should be made.

This is a cleanup PR so it should bring more clarity.

**Review by marc-hb — CHANGES_REQUESTED** [review]

File: tests/test_config.py (hunk 2 above)
Comment (pdgendt) [reply tests/test_config.py:33]: I was contemplating the same things, but tried to touch it "minimally" I'll try to clean this up so that it makes sense what is builtin and what not.

File: tests/test_config.py (hunk 6 above)
Comment (pdgendt) [reply tests/test_config.py]: Yeah, I tried to do it in some reasonable steps, but I think it can make it harder to follow.
I'll come back to this PR, but there's no urgency here, as I didn't intend on getting into 1.3.

File: tests/test_config.py (hunk 2 above)
Comment (pdgendt) [reply tests/test_config.py]: I've reverted to the original `configparser`.

File: tests/test_project.py (hunk 4 above)
Comment (pdgendt) [reply tests/test_project.py:1634]: Moved to a separate commit.

File: tests/test_config.py (hunk 6 above)
Comment (pdgendt) [reply tests/test_config.py]: Renamed as suggested, which is clearer.

File: tox.ini (hunk 5 above)
Comment (pdgendt) [reply tox.ini]: I've removed it from the `tox.ini` file and added it to the CI run.

File: tests/test_config.py [hunk 7]
```diff
@@ -30,9 +32,21 @@ def autouse_config_tmpdir(config_tmpdir):
 def cfg(f=ALL, topdir=None):
     # Load a fresh configuration object at the given level, and return it.
     cp = configparser.ConfigParser(allow_no_value=True)
-    config.read_config(configfile=f, config=cp, topdir=topdir)
+    cp.read(config._gather_configs(f, topdir), encoding='utf-8')
```
Comment (marc-hb) [inline tests/test_config.py:35]: After searching the code, I found that `_gather_configs()` is just an internal implementation detail of `read_config()` and used only by that deprecated function. This means `_gather_configs()` is effectively deprecated too. So I don't see what this particular line change brings except some confusion, sorry.

I think you should either "upgrade" this function to the newer API if possible, or not change it at all yet if not possible and/or too much work.

**Review by marc-hb — COMMENTED** [review]
Looks great except for one confusing, one-line change.

File: tests/test_config.py (hunk 7 above)
Comment (pdgendt) [reply tests/test_config.py]: I've wrapped the call in `with pytest.deprecated_call()` as it requires a lot more work in this case. Also added a `TODO` comment.

**Review by marc-hb — APPROVED** [review]
